src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs

```diff
@@ -310,12 +310,34 @@ internal class SqlServerFhirDataStore : IFhirDataStore, IProvideCapability
                     singleTransaction = true;
                 }
 
-                mergeWrappers.Add(new MergeResourceWrapper(resource, resourceExt.KeepHistory, hasVersionToCompare));
+                mergeWrappersWithVersions.Add((new MergeResourceWrapper(resource, resourceExt.KeepHistory, hasVersionToCompare), resourceExt.KeepVersion, int.Parse(resource.Version), existingVersion));
                 index++;
                 results.Add(identifier, new DataStoreOperationOutcome(new UpsertOutcome(resource, resource.Version == InitialVersion ? SaveOutcomeType.Created : SaveOutcomeType.Updated)));
             }
 
-            if (mergeWrappers.Count > 0) // Do not call DB with empty input
+            // Resources with input versions (keepVersion=true) might not have hasVersionToCompare set. Fix it here.
+            // Resources with keepVersion=true must be in separate call, and not mixed with keepVersion=false ones.
+            // Sort them in groups by resource id and order by version.
+            // In each group find the smallest version higher then existing
+            prevResourceId = string.Empty;
+            var notSetInResoureGroup = false;
+            foreach (var mergeWrapper in mergeWrappersWithVersions.Where(_ => _.KeepVersion && _.ExistingVersion != 0).OrderBy(_ => _.Wrapper.ResourceWrapper.ResourceId).ThenBy(_ => _.ResourceVersion))
+            {
+                if (prevResourceId != mergeWrapper.Wrapper.ResourceWrapper.ResourceId) // this should reset flag on each resource id group including first.
+                {
+                    notSetInResoureGroup = true;
+                }
+
+                prevResourceId = mergeWrapper.Wrapper.ResourceWrapper.ResourceId;
+
+                if (notSetInResoureGroup && mergeWrapper.ResourceVersion > mergeWrapper.ExistingVersion)
+                {
+                    mergeWrapper.Wrapper.HasVersionToCompare = true;
+                    notSetInResoureGroup = false;
+                }
+            }
```

Comment:
The linq version would look like this:

```csharp
foreach (var wrapperWithVersionToCompare in mergeWrappersWithVersions
             .Where(x => x.KeepVersion && x.ExistingVersion != 0)
             .GroupBy(key => key.Wrapper.ResourceWrapper.ResourceId)
             .Select(group => group.MinBy(y => y.ResourceVersion))
             .Where(x => x.ResourceVersion > x.ExistingVersion))
{
    wrapperWithVersionToCompare.Wrapper.HasVersionToCompare = true;
}
```

Reply:
Cool. It is more difficult for me to understand, so I want to stick to my current version.

Reply:
Sergey, I understand you have code that functions as intended. In this case though, .net has a `GroupBy` function that we can leverage.